### PR TITLE
PLT7931 - Makefile prereq check

### PR DESCRIPTION
**Pull request by AndersonWebStudio**
#### Summary
Adds a prerequisite check for Node, NPM, Docker, and Go

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-7931

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -400,7 +403,7 @@ run-client-fullmap: ## Runs the webapp with source code mapping (slower; better
 
 	cd $(BUILD_WEBAPP_DIR) && $(MAKE) run-fullmap
 
-run: run-server run-client ## Runs the server and webapp.
+run: check-prereqs run-server run-client ## Checks prerequisites, Runs the server and webapp.
```
Comment (ccbrown) [inline Makefile]: I don't think it's really necessary to add "Checks prerequisites" to the documentation. That's something that can happen transparently.

File: scripts/prereq-check.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+##check to see if node is installed
+type node >/dev/null 2>&1 || { echo >&2 "Mattermost requires NodeJS but it doesn't appear to be installed.  Aborting."; exit 1; }
+##check to see if it is at least the supported version
+if check_version $NODEVER $REQNODEVER; then
+    echo "Node minimum requirement met. Required: $REQNODEVER, Found: $NODEVER"
+else 
+    echo "WARNING! Mattermost did not find the minimum supported version of Node installed. Required: $REQNODEVER, Found: $NODEVER"
+    echo "We highly recommend stopping installation and updating dependencies before continuing"
+    read -p "Continue? (Y/n) " -n 1 -r
```
Comment (ccbrown) [inline scripts/prereq-check.sh]: So stopping is highly recommended, but continuing is the default?

File: scripts/prereq-check.sh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    echo
+    if [[ ! $REPLY =~ ^[Yy]$ ]]
```
Comment (ccbrown) [inline scripts/prereq-check.sh]: Is this right? Hitting enter for the default of "Y" to continue exits.

File: Makefile (hunk 1 above)
Comment (AndersonWebStudio) [reply Makefile]: I've removed that

File: scripts/prereq-check.sh (hunk 2 above)
Comment (AndersonWebStudio) [reply scripts/prereq-check.sh]: I adjusted this messaging to specifically ask for Y to move on, and note that it is not recommended to do so.

File: scripts/prereq-check.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    local dependency=$1 required_version=$2 installed_version=$3
+
+    type $dependency >/dev/null 2>&1 || { echo >&2 "Mattermost requires '$dependency' but it doesn't appear to be installed.  Aborting."; exit 1; }
+
+    if check_version $installed_version $required_version; then
+        echo "$dependency minimum requirement met. Required: $required_version, Found: $installed_version"
+    else 
+        echo "WARNING! Mattermost did not find the minimum supported version of '$dependency' installed. Required: $required_version, Found: $installed_version"
+        echo "We highly recommend stopping installation and updating dependencies before continuing"
+        read -p "Enter Y to continue anyway (not recommended).. " -n 1 -r
```
Comment (ccbrown) [inline scripts/prereq-check.sh]: ".."? Should probably be "." or "...".

**Comment by AndersonWebStudio** [comment]
@ccbrown I made some changes to this file to make it simpler, it now includes a check_prereq() function within the script to handle the duplicated pieces

**Comment by jwilander** [comment]
@AndersonWebStudio needs a rebase and then we can merge it in

**Pull request closed without merging**